### PR TITLE
Add DALi WidgetApplication creation guides

### PR DESCRIPTION
**Pull request by neostom432**
Now Widget feautres are available.
DALi user can create DALi widget application.

Signed-off-by: minho.sun <minho.sun@samsung.com>

### CODE REVIEW

**Comment by mijin85cho** [comment]
This PR is feedback-applied version of #248, so I assigned sabeena21 who is reviewing #248 as a reviewer.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+The table list the main function to manage widget instance:
+
+**Table: Main functions**
+
+| Function           | Description                                                                                   |
+|--------------------|-----------------------------------------------------------------------------------------------|
+| `OnCreate()`       | This function is called after the widget instance is created.                                 |
+| `OnTerminate()`    | This function is called after the widget instance is terminated.                              |
+| `OnPause()`        | This function is called when the widget is isinvisible.                                       |
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: This function is called when the widget is isinvisible. 
Please remove "is" from the sentence. Usage of is duplicate in the sentence.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+| `OnResume()`       | This function is called when the widget is visible.                                           |
+| `OnResize()`       | This function is called before the widget size is  changed.                                   |
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: "is  changed" 
please remove whitespace between is and changed in the sentence.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+   On starting, ensure that you must have background knowledge on [Handle/Body Pattern: Basic Way of Using DALi Objects](handle.md).
+
+   Additionally, for this pattern you need to construct two classes for custom widget:
+
+   * Handle class
+   * Body class
+
+   In this step, we will implement `Demo::SampleWidget` class and `Demo::Internal::SampleWidget` class as handle/body class.
+
+   1. Create handle class of custom widget by including <sample-widget.h> header file in your application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: should we use code tag for header file name?
`<sample-widget.h>` header file

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+     ~SampleWidget();
+     SampleWidget( const SampleWidget& sampleWidget );
+     SampleWidget& operator=( const SampleWidget& sampleWidget );
+   public:
+     explicit SampleWidget( Internal::SampleWidget* sampleWidget );
+   };
+   }
+   ```
+
+   2. Create body class of custom widget by using <sample-widget-impl.h> header file in your application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: should we use code tag for header file name?

`<sample-widget-impl.h>` header file

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+
+You can create a widget application with DALi to display small version of application on the home screen.
+
+To create a DALi widget application, you must:
+
+- Use the [Dali::WidgetApplication class](#widget-application-api).
+
+	> **Note**
+	> - The implementation of DALi widget application is similar to [DALi basic UI application](dali-basic-app.md), because the [Dali::WidgetApplication](../../api/mobile/latest/classDali_1_1WidgetApplication.html) class inherits from the [Dali::Application](../../api/mobile/latest/classDali_1_1Application.html) class.
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: [Dali::WidgetApplication](../../api/mobile/latest/classDali_1_1WidgetApplication.html)

change the path to:
[Dali::WidgetApplication]](../../api/mobile/4.0/classDali_1_1WidgetApplication.html)

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+       }
+
+     private:
+       WidgetApplication& mApplication;
+   };
+   ```
+
+## Related Information
+- Dependencies
+  - Tizen 4.0 and Higher
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: A Simple Question : I missed this point at previous review. Is this page is for both mobile and wearable?
If so, please write the dependencies for each profile like below
```
- Dependencies
  - Tizen 4.0  and Higher for Mobile
  - Tizen 4.0  and Higher for Wearable
```

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+   On starting, ensure that you must have background knowledge on [Handle/Body Pattern: Basic Way of Using DALi Objects](handle.md).
+
+   Additionally, for this pattern you need to construct two classes for custom widget:
+
+   * Handle class
+   * Body class
+
+   In this step, we will implement `Demo::SampleWidget` class and `Demo::Internal::SampleWidget` class as handle/body class.
+
+   1. Create handle class of custom widget by including sample-widget.h header file in your application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: please change sample-widget-impl.h to `<sample-widget-impl.h>`

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+     ~SampleWidget();
+     SampleWidget( const SampleWidget& sampleWidget );
+     SampleWidget& operator=( const SampleWidget& sampleWidget );
+   public:
+     explicit SampleWidget( Internal::SampleWidget* sampleWidget );
+   };
+   }
+   ```
+
+   2. Create body class of custom widget by using sample-widget-impl.h header file in your application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please change sample-widget-impl.h to `<sample-widget-impl.h>`

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+   void SampleWidget::OnUpdate( const std::string& contentInfo, int force )
+   {
+     DALI_LOG_ERROR("Widget instance updated\n");
+   }
+
+   } // Internal
+   } // Dali
+   ```
+
+   4. Include custom widget header and add Widget CreatingWidgetFunction() function for custom widget:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please change CreatingWidgetFunction()  to `CreatingWidgetFunction()`.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+        }
+
+      private:
+        WidgetApplication& mApplication;
+    }
+   ```
+
+3. Create own widget class
+
+   On starting, ensure that you must have background knowledge on [Handle/Body Pattern: Basic Way of Using DALi Objects](handle.md).
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Link `[Handle/Body Pattern: Basic Way of Using DALi Objects](handle.md)` is not working. Do you mean the file in `../../ui/dali/handle.md`?

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+   Additionally, for this pattern you need to construct two classes for custom widget:
+
+   * Handle class
+   * Body class
+
+   In this step, we will implement `Demo::SampleWidget` class and `Demo::Internal::SampleWidget` class as handle/body class.
+
+   1. Create handle class of custom widget by including `<sample-widget.h>` header file in your application:
+   ```
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please add 3 more spaces in front of the line from 116 to 137.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+     SampleWidget& operator=( const SampleWidget& sampleWidget );
+   public:
+     explicit SampleWidget( Internal::SampleWidget* sampleWidget );
+   };
+   }
+   ```
+
+   2. Create body class of custom widget by using `<sample-widget-impl.h>` header file in your application:
+
+   ```
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please add 3 more spaces in front of the line from 141 to 171.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+     SampleWidget(const SampleWidget&);
+     SampleWidget& operator=(SampleWidget&);
+   };
+   } // namespace Internal
+   } // namespace Dali
+   ```
+
+   3. Implement each lifecycle functions:
+
+   ```
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please add 3 more spaces in front of the line from 175 to 234.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+     DALI_LOG_ERROR("Widget instance updated\n");
+   }
+
+   } // Internal
+   } // Dali
+   ```
+
+   4. Include custom widget header and add Widget `CreatingWidgetFunction()` function for custom widget:
+
+   ```
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please add 3 more spaces in front of the line from 238 to 264.

**Comment by mijin85cho** [comment]
Thanks for the quick update, @neostom432.

Hello, @sabeena21, if you ok to go with the sentences in this PR, please approve this.
Then I'll merge this PR and apply the fix on Tizen Developer Site.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+   On starting, ensure that you must have background knowledge on [Handle/Body Pattern: Basic Way of Using DALi Objects](../../ui/dali/handle.md).
+
+   Additionally, for this pattern you need to construct two classes for custom widget:
+
+   * Handle class
+   * Body class
+
+   In this step, we will implement `Demo::SampleWidget` class and `Demo::Internal::SampleWidget` class as handle/body class.
+
+   1. Create handle class of custom widget by including `<sample-widget.h>` header file in your application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please rephrase as following:
To create handle class of custom widget include `<sample-widget.h>` header file in your application:

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+        ~SampleWidget();
+        SampleWidget( const SampleWidget& sampleWidget );
+        SampleWidget& operator=( const SampleWidget& sampleWidget );
+      public:
+        explicit SampleWidget( Internal::SampleWidget* sampleWidget );
+      };
+      }
+      ```
+
+   2. Create body class of custom widget by using `<sample-widget-impl.h>` header file in your application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please rephrase as the following:
To create body class of the custom widget use `<sample-widget-impl.h>` header file in your application:

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+      void SampleWidget::OnUpdate( const std::string& contentInfo, int force )
+      {
+        DALI_LOG_ERROR("Widget instance updated\n");
+      }
+
+      } // Internal
+      } // Dali
+      ```
+
+   4. Include custom widget header and add Widget `CreatingWidgetFunction()` function for custom widget:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please rephase as the following:

For custom widget include custom widget header. Also, add `Widget CreatingWidgetFunction()` function:

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+      protected:
+        // Undefined
+        SampleWidget(const SampleWidget&);
+        SampleWidget& operator=(SampleWidget&);
+      };
+      } // namespace Internal
+      } // namespace Dali
+      ```
+
+   3. Implement each lifecycle functions:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please rephrase as the following:
To implement each lifecycle function:

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+
+- Use the [Dali::WidgetApplication class](#widget-application-api).
+
+	> **Note**
+	> - The implementation of DALi widget application is similar to [DALi basic UI application](dali-basic-app.md), because the [Dali::WidgetApplication](../../../api/mobile/latest/classDali_1_1WidgetApplication.html) class inherits from the [Dali::Application](../../../api/mobile/latest/classDali_1_1Application.html) class.
+  > - Ensure you are familiar with the basic UI application details before tackling a widget application.
+
+- Ensure you start the event loop.
+
+- Implement your own widget class by inheriting the `Dali::Widget` class.
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please rephrase as the following:

To implement your own widget class inherit the `Dali::Widget` class.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+
+
+**Figure: DALi widget application**
+
+![DALi widget application](./media/dali_widget_application.png)
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: In our current checklist we have decided to add figure caption below the image.

Please add the **Figure: DALi widget application** caption below the image.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+
+For more information on basics of creating a DALi widget application, see [The steps to create a DALi widget application](#create).
+
+<a name="widget-application-api"></a>
+## Widget Application Class
+
+If you want to use DALi for your widget application, you must create the application with the `Dali::WidgetApplication` class. This class provides the means for initializing the resources required by DALi.
+
+The table list the main function you need in your widget application:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please change as following:
The table lists the main function you need in your widget application:

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+|--------------------------------------|------------------------------------------|
+| `mainloop()`                         | The mainloop() function is used to start the event loop. If you do not call the function and start the event loop, DALi cannot call any callback functions for application events.  |
+| `RegisterWidgetCreatingFunction()`   | This function is used to register create function to create widget instance. When widget viewer application requests, then the widget application creates widget instance using the function |
+
+<a name="widget-api"></a>
+## Widget Class
+
+`Dali::Widget` class provides interface for creating custom widget.
+
+The table list the main function to manage widget instance:
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please change as the following:
The table lists the main function to manage widget instance:

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+3. Create own widget class
+
+   On starting, ensure that you must have background knowledge on [Handle/Body Pattern: Basic Way of Using DALi Objects](../../ui/dali/handle.md).
+
+   Additionally, for this pattern you need to construct two classes for custom widget:
+
+   * Handle class
+   * Body class
+
+   In this step, we will implement `Demo::SampleWidget` class and `Demo::Internal::SampleWidget` class as handle/body class.
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please change as the following:

 In this step,  implement `Demo::SampleWidget` class and `Demo::Internal::SampleWidget` class as handle/body class.

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+
+
+You can create a widget application with DALi to display small version of application on the home screen.
+
+To create a DALi widget application, you must:
+
+- Use the [Dali::WidgetApplication class](#widget-application-api).
+
+	> **Note**
+	> - The implementation of DALi widget application is similar to [DALi basic UI application](dali-basic-app.md), because the [Dali::WidgetApplication](../../../api/mobile/latest/classDali_1_1WidgetApplication.html) class inherits from the [Dali::Application](../../../api/mobile/latest/classDali_1_1Application.html) class.
```
[the remark this answers is not in the document]
Comment (neostom432) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: It worked on my local...
I don't know why it didn't work.

The tree of directory is below.
- native
  - api
    - mobile
      - latest
        - classDali_1_1WidgetApplication.html
  - guides
    - app-management
      - ui-app
        - dali-widget-app.md 

To reach classDali_1_1WidgetApplication.html, we should move to parent third time( ../../../ ) and enter to api/mobile/latest.

So the path will be ../../../api/mobile/latest/classDali_1_1WidgetApplication.html

I think it is kind of bug....

[the remark this answers is not in the document]
Comment (sabeena21) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: @neostom432 ok, so the path of the URL will  to be changed or will it be the same?

[the remark this answers is not in the document]
Comment (mijin85cho) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Hello.

I checked the updated URL and it is working fine.

For now they are linking to below.

- `https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/latest/classDali_1_1WidgetApplication.html`

- `https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/latest/classDali_1_1Application.html`

As symbolic link is not working on github now, while checking the link, 
we should change the `latest` to the version we want to check manually on Address bar.
(The working of symbolic link will be updated on github later.)

For now, the latest version is 4.0. So we should check whether below links are working.

- `https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/4.0/classDali_1_1WidgetApplication.html`

- `https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/4.0/classDali_1_1Application.html`

As you can see, the URLs are updated and working well.

Regards

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+  > - Ensure you are familiar with the basic UI application details before tackling a widget application.
+
+- Ensure you start the event loop.
+
+- To implement your own widget class inherit the `Dali::Widget class`.
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Question : Is "`Dali::Widget class`" OK? 
Wasn't it "`Dali::Widget` class" ?

Comment (neostom432) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Done

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (sabeena21) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Hello @mijin85cho,

The following two links are working when I have pasted it on the browser:

- https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/4.0/classDali_1_1WidgetApplication.html

- https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/4.0/classDali_1_1Application.html

The following two links are not working when I have pasted on the browser:
- https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/latest/classDali_1_1WidgetApplication.html

- https://github.com/neostom432/tizen-docs/blob/e57c41e3ef12007a9b2c095ab1f458bd5843d0f1/docs/application/native/api/mobile/latest/classDali_1_1Application.html

Let me confirm that the non working link will be updated to working link later?

[the remark this answers is not in the document]
Comment (mijin85cho) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Hello, @sabeena21 

Yes, it is expected that this repository will support **symbolic link** later.

----------
Also please refer https://en.wikipedia.org/wiki/Symbolic_link
```
In computing, a symbolic link (also symlink or soft link) is a term for any file 
that contains a reference to another file or directory 
in the form of an absolute or relative path and that affects pathname resolution.
```

As you can see, `tizen-docs/docs/application/native/api/mobile/latest` is a symbolic link.
When you open it on master branch, you'll see only "4.0/" is written it.
That is why we change "latest" in the URL to "4.0"

As every links from guide to API reference is using symbolic link, 
for the futural review, please consider that symbolic link is not working on github for now.
It means you should check every url to api reference with changing to "4.0" or "5.0" for a while...

Regards
mijin

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+        {
+          // Register widget creating function here.
+        }
+
+      private:
+        WidgetApplication& mApplication;
+    }
+   ```
+
+3. Create own widget class
```
Comment (sabeena21) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Please can you change the sentence as:
Create a widget class.

Comment (neostom432) [reply docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: Done

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+**Figure: DALi widget application**
+
+For more information on basics of creating a DALi widget application, see [The steps to create a DALi widget application](#create).
+
+<a name="widget-application-api"></a>
+## Widget Application Class
+
+If you want to use DALi for your widget application, you must create the application with the `Dali::WidgetApplication` class. This class provides the means for initializing the resources required by DALi.
+
+The table lists the main function to manage widget instance:
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: "main function", of "main function**s**"?

File: docs/application/native/guides/app-management/ui-app/dali-widget-app.md [hunk 28]
```diff
@@ -0,0 +1,268 @@
...
+|--------------------------------------|------------------------------------------|
+| `mainloop()`                         | The mainloop() function is used to start the event loop. If you do not call the function and start the event loop, DALi cannot call any callback functions for application events.  |
+| `RegisterWidgetCreatingFunction()`   | This function is used to register create function to create widget instance. When widget viewer application requests, then the widget application creates widget instance using the function |
+
+<a name="widget-api"></a>
+## Widget Class
+
+`Dali::Widget` class provides interface for creating custom widget.
+
+The table list the main function to manage widget instance:
```
Comment (mijin85cho) [inline docs/application/native/guides/app-management/ui-app/dali-widget-app.md]: 1. On line 29, "lists" is used in the third person singular. How about here? Isn't this "list" should be "lists", too?

2. "main function", of "main functions"?

**Comment by mijin85cho** [comment]
Hello, @neostom432 
Would you please update this PR as sabeena21 requested?

**Comment by neostom432** [comment]
@mijin85cho , @sabeena21 
Updated. Please check.

**Comment by mijin85cho** [comment]
Thanks for the update & review.

**Comment by mijin80lee** [comment]
TD Sync Completed.

**Pull request merged**